File: Python3/0031_next_permutation.py
# https://leetcode-cn.com/problems/next-permutation/

class Solution:
    def nextPermutation(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            None
        """
        
        """
        Do not return anything, modify nums in-place instead.
        """
        
        n = len(nums)
        i, j = n-2, n-1
        flag = False
        while i >= 0 :
            if nums[i] < nums[j]:
                flag = True
                break
            else:
                i = i - 1
                j = j - 1
        
        if not flag:
            nums.sort()
        else:
            k = n - 1
            while k >= j:
                if nums[k] > nums[i]:
                    nums[i], nums[k] = nums[k], nums[i]
                    # Sort nums[j:n].
                    for idx in range((n-j)//2):
                        nums[j+idx], nums[n-1-idx] = nums[n-1-idx], nums[j+idx]
                    break
                else:
                    k = k - 1

File: Cpp/0031_next_permutation.cpp
// https://leetcode.cn/problems/next-permutation/

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size(), idx = n - 1;
        while (idx > 0) {
            if (nums[idx-1] < nums[idx]) {
                int tmp = nums[idx-1];
                // Find the smallest values that is greater than nums[idx-1]
                // and that is on the right of nums[idx-1].
                for (int i = n - 1; i >= idx; --i) {
                    if (tmp < nums[i]) {
                        nums[idx-1] = nums[i];
                        nums[i] = tmp;
                        break;
                    }
                }
                break;
            }

            idx = idx - 1;
        }

        // Reverse.
        for (int i = 0; i < (n-idx)/2; ++i) {
            int tmp = nums[idx+i];
            nums[idx+i] = nums[n-1-i];
            nums[n-1-i] = tmp;
        }

    }
};